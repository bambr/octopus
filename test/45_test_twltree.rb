#!/usr/bin/ruby

$: << File.dirname($0) + '/lib'
require 'run_env'

class NormalEnv < RunEnv
	def config
		super + <<EOD

object_space[0].index[1].type = "TREE"
object_space[0].index[1].unique = 1
object_space[0].index[1].key_field[0].fieldno = 1
object_space[0].index[1].key_field[0].type = "STR"
EOD
	end
end

NormalEnv.env_eval do
	start

	c = connect

	30.times do |i|
		c.insert [i + 0, "abc#{i} + 0"]
		c.insert [i + 1, "abc#{i} + 1"]
		c.insert [i + 2, "abc#{i} + 2"]
		c.select i + 0, i + 1, i + 2

		snapshot
		sleep 0.1
		stop
		sleep 0.1
		start
		sleep 0.1
		c.reconnect

		c.select i + 0, i + 1, i + 2
		c.delete i + 0
		c.select i + 0, i + 1, i + 2
		c.delete i + 1
		c.select i + 0, i + 1, i + 2
		c.delete i + 2
		c.select i + 0, i + 1, i + 2
	end

	3000.times do |i|
		c.select i
		c.insert [i, "abc#{i}"]
		c.select i
	end

	snapshot
	sleep 0.1
	stop
	sleep 0.1
	start
	sleep 0.1
	c.reconnect

	3000.times do |i|
		c.select i
		c.delete i
		c.select i
	end
end
