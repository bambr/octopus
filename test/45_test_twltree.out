# box.insert([0, "abc0 + 0", 0])
1

# box.insert([1, "abc0 + 1", 1])
1

# box.insert([2, "abc0 + 2", 2])
1

# box.select(0, 1, 2)
[["\x00\x00\x00\x00", "abc0 + 0", "\x00\x00\x00\x00"], ["\x01\x00\x00\x00", "abc0 + 1", "\x01\x00\x00\x00"], ["\x02\x00\x00\x00", "abc0 + 2", "\x02\x00\x00\x00"]]

# box.select(["abc0 + 0"], {:index=>1, :limit=>1})
[["\x00\x00\x00\x00", "abc0 + 0", "\x00\x00\x00\x00"]]

# box.select(["abc0 + 0"], {:index=>2, :limit=>1})
[["\x00\x00\x00\x00", "abc0 + 0", "\x00\x00\x00\x00"]]

# box.select(["abc0 + 0", 0], {:index=>2, :limit=>1})
[["\x00\x00\x00\x00", "abc0 + 0", "\x00\x00\x00\x00"]]

# box.select(["abc0 + 1"], {:index=>1, :limit=>1})
[["\x01\x00\x00\x00", "abc0 + 1", "\x01\x00\x00\x00"]]

# box.select(["abc0 + 1"], {:index=>2, :limit=>1})
[["\x01\x00\x00\x00", "abc0 + 1", "\x01\x00\x00\x00"]]

# box.select(["abc0 + 1", 1], {:index=>2, :limit=>1})
[["\x01\x00\x00\x00", "abc0 + 1", "\x01\x00\x00\x00"]]

# box.select(["abc0 + 2"], {:index=>1, :limit=>1})
[["\x02\x00\x00\x00", "abc0 + 2", "\x02\x00\x00\x00"]]

# box.select(["abc0 + 2"], {:index=>2, :limit=>1})
[["\x02\x00\x00\x00", "abc0 + 2", "\x02\x00\x00\x00"]]

# box.select(["abc0 + 2", 2], {:index=>2, :limit=>1})
[["\x02\x00\x00\x00", "abc0 + 2", "\x02\x00\x00\x00"]]

# box.select(0, 1, 2)
[["\x00\x00\x00\x00", "abc0 + 0", "\x00\x00\x00\x00"], ["\x01\x00\x00\x00", "abc0 + 1", "\x01\x00\x00\x00"], ["\x02\x00\x00\x00", "abc0 + 2", "\x02\x00\x00\x00"]]

# box.delete(0)
1

# box.select(0, 1, 2)
[["\x01\x00\x00\x00", "abc0 + 1", "\x01\x00\x00\x00"], ["\x02\x00\x00\x00", "abc0 + 2", "\x02\x00\x00\x00"]]

# box.delete(1)
1

# box.select(0, 1, 2)
[["\x02\x00\x00\x00", "abc0 + 2", "\x02\x00\x00\x00"]]

# box.delete(2)
1

# box.select(0, 1, 2)
[]

# box.insert([1, "abc1 + 0", 1])
1

# box.insert([2, "abc1 + 1", 2])
1

# box.insert([3, "abc1 + 2", 3])
1

# box.select(1, 2, 3)
[["\x01\x00\x00\x00", "abc1 + 0", "\x01\x00\x00\x00"], ["\x02\x00\x00\x00", "abc1 + 1", "\x02\x00\x00\x00"], ["\x03\x00\x00\x00", "abc1 + 2", "\x03\x00\x00\x00"]]

# box.select(["abc1 + 0"], {:index=>1, :limit=>1})
[["\x01\x00\x00\x00", "abc1 + 0", "\x01\x00\x00\x00"]]

# box.select(["abc1 + 0"], {:index=>2, :limit=>1})
[["\x01\x00\x00\x00", "abc1 + 0", "\x01\x00\x00\x00"]]

# box.select(["abc1 + 0", 1], {:index=>2, :limit=>1})
[["\x01\x00\x00\x00", "abc1 + 0", "\x01\x00\x00\x00"]]

# box.select(["abc1 + 1"], {:index=>1, :limit=>1})
[["\x02\x00\x00\x00", "abc1 + 1", "\x02\x00\x00\x00"]]

# box.select(["abc1 + 1"], {:index=>2, :limit=>1})
[["\x02\x00\x00\x00", "abc1 + 1", "\x02\x00\x00\x00"]]

# box.select(["abc1 + 1", 2], {:index=>2, :limit=>1})
[["\x02\x00\x00\x00", "abc1 + 1", "\x02\x00\x00\x00"]]

# box.select(["abc1 + 2"], {:index=>1, :limit=>1})
[["\x03\x00\x00\x00", "abc1 + 2", "\x03\x00\x00\x00"]]

# box.select(["abc1 + 2"], {:index=>2, :limit=>1})
[["\x03\x00\x00\x00", "abc1 + 2", "\x03\x00\x00\x00"]]

# box.select(["abc1 + 2", 3], {:index=>2, :limit=>1})
[["\x03\x00\x00\x00", "abc1 + 2", "\x03\x00\x00\x00"]]

# box.select(1, 2, 3)
[["\x01\x00\x00\x00", "abc1 + 0", "\x01\x00\x00\x00"], ["\x02\x00\x00\x00", "abc1 + 1", "\x02\x00\x00\x00"], ["\x03\x00\x00\x00", "abc1 + 2", "\x03\x00\x00\x00"]]

# box.delete(1)
1

# box.select(1, 2, 3)
[["\x02\x00\x00\x00", "abc1 + 1", "\x02\x00\x00\x00"], ["\x03\x00\x00\x00", "abc1 + 2", "\x03\x00\x00\x00"]]

# box.delete(2)
1

# box.select(1, 2, 3)
[["\x03\x00\x00\x00", "abc1 + 2", "\x03\x00\x00\x00"]]

# box.delete(3)
1

# box.select(1, 2, 3)
[]

# box.insert([2, "abc2 + 0", 2])
1

# box.insert([3, "abc2 + 1", 3])
1

# box.insert([4, "abc2 + 2", 4])
1

# box.select(2, 3, 4)
[["\x02\x00\x00\x00", "abc2 + 0", "\x02\x00\x00\x00"], ["\x03\x00\x00\x00", "abc2 + 1", "\x03\x00\x00\x00"], ["\x04\x00\x00\x00", "abc2 + 2", "\x04\x00\x00\x00"]]

# box.select(["abc2 + 0"], {:index=>1, :limit=>1})
[["\x02\x00\x00\x00", "abc2 + 0", "\x02\x00\x00\x00"]]

# box.select(["abc2 + 0"], {:index=>2, :limit=>1})
[["\x02\x00\x00\x00", "abc2 + 0", "\x02\x00\x00\x00"]]

# box.select(["abc2 + 0", 2], {:index=>2, :limit=>1})
[["\x02\x00\x00\x00", "abc2 + 0", "\x02\x00\x00\x00"]]

# box.select(["abc2 + 1"], {:index=>1, :limit=>1})
[["\x03\x00\x00\x00", "abc2 + 1", "\x03\x00\x00\x00"]]

# box.select(["abc2 + 1"], {:index=>2, :limit=>1})
[["\x03\x00\x00\x00", "abc2 + 1", "\x03\x00\x00\x00"]]

# box.select(["abc2 + 1", 3], {:index=>2, :limit=>1})
[["\x03\x00\x00\x00", "abc2 + 1", "\x03\x00\x00\x00"]]

# box.select(["abc2 + 2"], {:index=>1, :limit=>1})
[["\x04\x00\x00\x00", "abc2 + 2", "\x04\x00\x00\x00"]]

# box.select(["abc2 + 2"], {:index=>2, :limit=>1})
[["\x04\x00\x00\x00", "abc2 + 2", "\x04\x00\x00\x00"]]

# box.select(["abc2 + 2", 4], {:index=>2, :limit=>1})
[["\x04\x00\x00\x00", "abc2 + 2", "\x04\x00\x00\x00"]]

# box.select(2, 3, 4)
[["\x02\x00\x00\x00", "abc2 + 0", "\x02\x00\x00\x00"], ["\x03\x00\x00\x00", "abc2 + 1", "\x03\x00\x00\x00"], ["\x04\x00\x00\x00", "abc2 + 2", "\x04\x00\x00\x00"]]

# box.delete(2)
1

# box.select(2, 3, 4)
[["\x03\x00\x00\x00", "abc2 + 1", "\x03\x00\x00\x00"], ["\x04\x00\x00\x00", "abc2 + 2", "\x04\x00\x00\x00"]]

# box.delete(3)
1

# box.select(2, 3, 4)
[["\x04\x00\x00\x00", "abc2 + 2", "\x04\x00\x00\x00"]]

# box.delete(4)
1

# box.select(2, 3, 4)
[]

# box.insert([3, "abc3 + 0", 3])
1

# box.insert([4, "abc3 + 1", 4])
1

# box.insert([5, "abc3 + 2", 5])
1

# box.select(3, 4, 5)
[["\x03\x00\x00\x00", "abc3 + 0", "\x03\x00\x00\x00"], ["\x04\x00\x00\x00", "abc3 + 1", "\x04\x00\x00\x00"], ["\x05\x00\x00\x00", "abc3 + 2", "\x05\x00\x00\x00"]]

# box.select(["abc3 + 0"], {:index=>1, :limit=>1})
[["\x03\x00\x00\x00", "abc3 + 0", "\x03\x00\x00\x00"]]

# box.select(["abc3 + 0"], {:index=>2, :limit=>1})
[["\x03\x00\x00\x00", "abc3 + 0", "\x03\x00\x00\x00"]]

# box.select(["abc3 + 0", 3], {:index=>2, :limit=>1})
[["\x03\x00\x00\x00", "abc3 + 0", "\x03\x00\x00\x00"]]

# box.select(["abc3 + 1"], {:index=>1, :limit=>1})
[["\x04\x00\x00\x00", "abc3 + 1", "\x04\x00\x00\x00"]]

# box.select(["abc3 + 1"], {:index=>2, :limit=>1})
[["\x04\x00\x00\x00", "abc3 + 1", "\x04\x00\x00\x00"]]

# box.select(["abc3 + 1", 4], {:index=>2, :limit=>1})
[["\x04\x00\x00\x00", "abc3 + 1", "\x04\x00\x00\x00"]]

# box.select(["abc3 + 2"], {:index=>1, :limit=>1})
[["\x05\x00\x00\x00", "abc3 + 2", "\x05\x00\x00\x00"]]

# box.select(["abc3 + 2"], {:index=>2, :limit=>1})
[["\x05\x00\x00\x00", "abc3 + 2", "\x05\x00\x00\x00"]]

# box.select(["abc3 + 2", 5], {:index=>2, :limit=>1})
[["\x05\x00\x00\x00", "abc3 + 2", "\x05\x00\x00\x00"]]

# box.select(3, 4, 5)
[["\x03\x00\x00\x00", "abc3 + 0", "\x03\x00\x00\x00"], ["\x04\x00\x00\x00", "abc3 + 1", "\x04\x00\x00\x00"], ["\x05\x00\x00\x00", "abc3 + 2", "\x05\x00\x00\x00"]]

# box.delete(3)
1

# box.select(3, 4, 5)
[["\x04\x00\x00\x00", "abc3 + 1", "\x04\x00\x00\x00"], ["\x05\x00\x00\x00", "abc3 + 2", "\x05\x00\x00\x00"]]

# box.delete(4)
1

# box.select(3, 4, 5)
[["\x05\x00\x00\x00", "abc3 + 2", "\x05\x00\x00\x00"]]

# box.delete(5)
1

# box.select(3, 4, 5)
[]

# box.insert([4, "abc4 + 0", 4])
1

# box.insert([5, "abc4 + 1", 5])
1

# box.insert([6, "abc4 + 2", 6])
1

# box.select(4, 5, 6)
[["\x04\x00\x00\x00", "abc4 + 0", "\x04\x00\x00\x00"], ["\x05\x00\x00\x00", "abc4 + 1", "\x05\x00\x00\x00"], ["\x06\x00\x00\x00", "abc4 + 2", "\x06\x00\x00\x00"]]

# box.select(["abc4 + 0"], {:index=>1, :limit=>1})
[["\x04\x00\x00\x00", "abc4 + 0", "\x04\x00\x00\x00"]]

# box.select(["abc4 + 0"], {:index=>2, :limit=>1})
[["\x04\x00\x00\x00", "abc4 + 0", "\x04\x00\x00\x00"]]

# box.select(["abc4 + 0", 4], {:index=>2, :limit=>1})
[["\x04\x00\x00\x00", "abc4 + 0", "\x04\x00\x00\x00"]]

# box.select(["abc4 + 1"], {:index=>1, :limit=>1})
[["\x05\x00\x00\x00", "abc4 + 1", "\x05\x00\x00\x00"]]

# box.select(["abc4 + 1"], {:index=>2, :limit=>1})
[["\x05\x00\x00\x00", "abc4 + 1", "\x05\x00\x00\x00"]]

# box.select(["abc4 + 1", 5], {:index=>2, :limit=>1})
[["\x05\x00\x00\x00", "abc4 + 1", "\x05\x00\x00\x00"]]

# box.select(["abc4 + 2"], {:index=>1, :limit=>1})
[["\x06\x00\x00\x00", "abc4 + 2", "\x06\x00\x00\x00"]]

# box.select(["abc4 + 2"], {:index=>2, :limit=>1})
[["\x06\x00\x00\x00", "abc4 + 2", "\x06\x00\x00\x00"]]

# box.select(["abc4 + 2", 6], {:index=>2, :limit=>1})
[["\x06\x00\x00\x00", "abc4 + 2", "\x06\x00\x00\x00"]]

# box.select(4, 5, 6)
[["\x04\x00\x00\x00", "abc4 + 0", "\x04\x00\x00\x00"], ["\x05\x00\x00\x00", "abc4 + 1", "\x05\x00\x00\x00"], ["\x06\x00\x00\x00", "abc4 + 2", "\x06\x00\x00\x00"]]

# box.delete(4)
1

# box.select(4, 5, 6)
[["\x05\x00\x00\x00", "abc4 + 1", "\x05\x00\x00\x00"], ["\x06\x00\x00\x00", "abc4 + 2", "\x06\x00\x00\x00"]]

# box.delete(5)
1

# box.select(4, 5, 6)
[["\x06\x00\x00\x00", "abc4 + 2", "\x06\x00\x00\x00"]]

# box.delete(6)
1

# box.select(4, 5, 6)
[]

# box.insert([5, "abc5 + 0", 5])
1

# box.insert([6, "abc5 + 1", 6])
1

# box.insert([7, "abc5 + 2", 7])
1

# box.select(5, 6, 7)
[["\x05\x00\x00\x00", "abc5 + 0", "\x05\x00\x00\x00"], ["\x06\x00\x00\x00", "abc5 + 1", "\x06\x00\x00\x00"], ["\a\x00\x00\x00", "abc5 + 2", "\a\x00\x00\x00"]]

# box.select(["abc5 + 0"], {:index=>1, :limit=>1})
[["\x05\x00\x00\x00", "abc5 + 0", "\x05\x00\x00\x00"]]

# box.select(["abc5 + 0"], {:index=>2, :limit=>1})
[["\x05\x00\x00\x00", "abc5 + 0", "\x05\x00\x00\x00"]]

# box.select(["abc5 + 0", 5], {:index=>2, :limit=>1})
[["\x05\x00\x00\x00", "abc5 + 0", "\x05\x00\x00\x00"]]

# box.select(["abc5 + 1"], {:index=>1, :limit=>1})
[["\x06\x00\x00\x00", "abc5 + 1", "\x06\x00\x00\x00"]]

# box.select(["abc5 + 1"], {:index=>2, :limit=>1})
[["\x06\x00\x00\x00", "abc5 + 1", "\x06\x00\x00\x00"]]

# box.select(["abc5 + 1", 6], {:index=>2, :limit=>1})
[["\x06\x00\x00\x00", "abc5 + 1", "\x06\x00\x00\x00"]]

# box.select(["abc5 + 2"], {:index=>1, :limit=>1})
[["\a\x00\x00\x00", "abc5 + 2", "\a\x00\x00\x00"]]

# box.select(["abc5 + 2"], {:index=>2, :limit=>1})
[["\a\x00\x00\x00", "abc5 + 2", "\a\x00\x00\x00"]]

# box.select(["abc5 + 2", 7], {:index=>2, :limit=>1})
[["\a\x00\x00\x00", "abc5 + 2", "\a\x00\x00\x00"]]

# box.select(5, 6, 7)
[["\x05\x00\x00\x00", "abc5 + 0", "\x05\x00\x00\x00"], ["\x06\x00\x00\x00", "abc5 + 1", "\x06\x00\x00\x00"], ["\a\x00\x00\x00", "abc5 + 2", "\a\x00\x00\x00"]]

# box.delete(5)
1

# box.select(5, 6, 7)
[["\x06\x00\x00\x00", "abc5 + 1", "\x06\x00\x00\x00"], ["\a\x00\x00\x00", "abc5 + 2", "\a\x00\x00\x00"]]

# box.delete(6)
1

# box.select(5, 6, 7)
[["\a\x00\x00\x00", "abc5 + 2", "\a\x00\x00\x00"]]

# box.delete(7)
1

# box.select(5, 6, 7)
[]

# box.insert([6, "abc6 + 0", 6])
1

# box.insert([7, "abc6 + 1", 7])
1

# box.insert([8, "abc6 + 2", 8])
1

# box.select(6, 7, 8)
[["\x06\x00\x00\x00", "abc6 + 0", "\x06\x00\x00\x00"], ["\a\x00\x00\x00", "abc6 + 1", "\a\x00\x00\x00"], ["\b\x00\x00\x00", "abc6 + 2", "\b\x00\x00\x00"]]

# box.select(["abc6 + 0"], {:index=>1, :limit=>1})
[["\x06\x00\x00\x00", "abc6 + 0", "\x06\x00\x00\x00"]]

# box.select(["abc6 + 0"], {:index=>2, :limit=>1})
[["\x06\x00\x00\x00", "abc6 + 0", "\x06\x00\x00\x00"]]

# box.select(["abc6 + 0", 6], {:index=>2, :limit=>1})
[["\x06\x00\x00\x00", "abc6 + 0", "\x06\x00\x00\x00"]]

# box.select(["abc6 + 1"], {:index=>1, :limit=>1})
[["\a\x00\x00\x00", "abc6 + 1", "\a\x00\x00\x00"]]

# box.select(["abc6 + 1"], {:index=>2, :limit=>1})
[["\a\x00\x00\x00", "abc6 + 1", "\a\x00\x00\x00"]]

# box.select(["abc6 + 1", 7], {:index=>2, :limit=>1})
[["\a\x00\x00\x00", "abc6 + 1", "\a\x00\x00\x00"]]

# box.select(["abc6 + 2"], {:index=>1, :limit=>1})
[["\b\x00\x00\x00", "abc6 + 2", "\b\x00\x00\x00"]]

# box.select(["abc6 + 2"], {:index=>2, :limit=>1})
[["\b\x00\x00\x00", "abc6 + 2", "\b\x00\x00\x00"]]

# box.select(["abc6 + 2", 8], {:index=>2, :limit=>1})
[["\b\x00\x00\x00", "abc6 + 2", "\b\x00\x00\x00"]]

# box.select(6, 7, 8)
[["\x06\x00\x00\x00", "abc6 + 0", "\x06\x00\x00\x00"], ["\a\x00\x00\x00", "abc6 + 1", "\a\x00\x00\x00"], ["\b\x00\x00\x00", "abc6 + 2", "\b\x00\x00\x00"]]

# box.delete(6)
1

# box.select(6, 7, 8)
[["\a\x00\x00\x00", "abc6 + 1", "\a\x00\x00\x00"], ["\b\x00\x00\x00", "abc6 + 2", "\b\x00\x00\x00"]]

# box.delete(7)
1

# box.select(6, 7, 8)
[["\b\x00\x00\x00", "abc6 + 2", "\b\x00\x00\x00"]]

# box.delete(8)
1

# box.select(6, 7, 8)
[]

# box.insert([7, "abc7 + 0", 7])
1

# box.insert([8, "abc7 + 1", 8])
1

# box.insert([9, "abc7 + 2", 9])
1

# box.select(7, 8, 9)
[["\a\x00\x00\x00", "abc7 + 0", "\a\x00\x00\x00"], ["\b\x00\x00\x00", "abc7 + 1", "\b\x00\x00\x00"], ["\t\x00\x00\x00", "abc7 + 2", "\t\x00\x00\x00"]]

# box.select(["abc7 + 0"], {:index=>1, :limit=>1})
[["\a\x00\x00\x00", "abc7 + 0", "\a\x00\x00\x00"]]

# box.select(["abc7 + 0"], {:index=>2, :limit=>1})
[["\a\x00\x00\x00", "abc7 + 0", "\a\x00\x00\x00"]]

# box.select(["abc7 + 0", 7], {:index=>2, :limit=>1})
[["\a\x00\x00\x00", "abc7 + 0", "\a\x00\x00\x00"]]

# box.select(["abc7 + 1"], {:index=>1, :limit=>1})
[["\b\x00\x00\x00", "abc7 + 1", "\b\x00\x00\x00"]]

# box.select(["abc7 + 1"], {:index=>2, :limit=>1})
[["\b\x00\x00\x00", "abc7 + 1", "\b\x00\x00\x00"]]

# box.select(["abc7 + 1", 8], {:index=>2, :limit=>1})
[["\b\x00\x00\x00", "abc7 + 1", "\b\x00\x00\x00"]]

# box.select(["abc7 + 2"], {:index=>1, :limit=>1})
[["\t\x00\x00\x00", "abc7 + 2", "\t\x00\x00\x00"]]

# box.select(["abc7 + 2"], {:index=>2, :limit=>1})
[["\t\x00\x00\x00", "abc7 + 2", "\t\x00\x00\x00"]]

# box.select(["abc7 + 2", 9], {:index=>2, :limit=>1})
[["\t\x00\x00\x00", "abc7 + 2", "\t\x00\x00\x00"]]

# box.select(7, 8, 9)
[["\a\x00\x00\x00", "abc7 + 0", "\a\x00\x00\x00"], ["\b\x00\x00\x00", "abc7 + 1", "\b\x00\x00\x00"], ["\t\x00\x00\x00", "abc7 + 2", "\t\x00\x00\x00"]]

# box.delete(7)
1

# box.select(7, 8, 9)
[["\b\x00\x00\x00", "abc7 + 1", "\b\x00\x00\x00"], ["\t\x00\x00\x00", "abc7 + 2", "\t\x00\x00\x00"]]

# box.delete(8)
1

# box.select(7, 8, 9)
[["\t\x00\x00\x00", "abc7 + 2", "\t\x00\x00\x00"]]

# box.delete(9)
1

# box.select(7, 8, 9)
[]

# box.insert([8, "abc8 + 0", 8])
1

# box.insert([9, "abc8 + 1", 9])
1

# box.insert([10, "abc8 + 2", 10])
1

# box.select(8, 9, 10)
[["\b\x00\x00\x00", "abc8 + 0", "\b\x00\x00\x00"], ["\t\x00\x00\x00", "abc8 + 1", "\t\x00\x00\x00"], ["\n\x00\x00\x00", "abc8 + 2", "\n\x00\x00\x00"]]

# box.select(["abc8 + 0"], {:index=>1, :limit=>1})
[["\b\x00\x00\x00", "abc8 + 0", "\b\x00\x00\x00"]]

# box.select(["abc8 + 0"], {:index=>2, :limit=>1})
[["\b\x00\x00\x00", "abc8 + 0", "\b\x00\x00\x00"]]

# box.select(["abc8 + 0", 8], {:index=>2, :limit=>1})
[["\b\x00\x00\x00", "abc8 + 0", "\b\x00\x00\x00"]]

# box.select(["abc8 + 1"], {:index=>1, :limit=>1})
[["\t\x00\x00\x00", "abc8 + 1", "\t\x00\x00\x00"]]

# box.select(["abc8 + 1"], {:index=>2, :limit=>1})
[["\t\x00\x00\x00", "abc8 + 1", "\t\x00\x00\x00"]]

# box.select(["abc8 + 1", 9], {:index=>2, :limit=>1})
[["\t\x00\x00\x00", "abc8 + 1", "\t\x00\x00\x00"]]

# box.select(["abc8 + 2"], {:index=>1, :limit=>1})
[["\n\x00\x00\x00", "abc8 + 2", "\n\x00\x00\x00"]]

# box.select(["abc8 + 2"], {:index=>2, :limit=>1})
[["\n\x00\x00\x00", "abc8 + 2", "\n\x00\x00\x00"]]

# box.select(["abc8 + 2", 10], {:index=>2, :limit=>1})
[["\n\x00\x00\x00", "abc8 + 2", "\n\x00\x00\x00"]]

# box.select(8, 9, 10)
[["\b\x00\x00\x00", "abc8 + 0", "\b\x00\x00\x00"], ["\t\x00\x00\x00", "abc8 + 1", "\t\x00\x00\x00"], ["\n\x00\x00\x00", "abc8 + 2", "\n\x00\x00\x00"]]

# box.delete(8)
1

# box.select(8, 9, 10)
[["\t\x00\x00\x00", "abc8 + 1", "\t\x00\x00\x00"], ["\n\x00\x00\x00", "abc8 + 2", "\n\x00\x00\x00"]]

# box.delete(9)
1

# box.select(8, 9, 10)
[["\n\x00\x00\x00", "abc8 + 2", "\n\x00\x00\x00"]]

# box.delete(10)
1

# box.select(8, 9, 10)
[]

# box.insert([9, "abc9 + 0", 9])
1

# box.insert([10, "abc9 + 1", 10])
1

# box.insert([11, "abc9 + 2", 11])
1

# box.select(9, 10, 11)
[["\t\x00\x00\x00", "abc9 + 0", "\t\x00\x00\x00"], ["\n\x00\x00\x00", "abc9 + 1", "\n\x00\x00\x00"], ["\v\x00\x00\x00", "abc9 + 2", "\v\x00\x00\x00"]]

# box.select(["abc9 + 0"], {:index=>1, :limit=>1})
[["\t\x00\x00\x00", "abc9 + 0", "\t\x00\x00\x00"]]

# box.select(["abc9 + 0"], {:index=>2, :limit=>1})
[["\t\x00\x00\x00", "abc9 + 0", "\t\x00\x00\x00"]]

# box.select(["abc9 + 0", 9], {:index=>2, :limit=>1})
[["\t\x00\x00\x00", "abc9 + 0", "\t\x00\x00\x00"]]

# box.select(["abc9 + 1"], {:index=>1, :limit=>1})
[["\n\x00\x00\x00", "abc9 + 1", "\n\x00\x00\x00"]]

# box.select(["abc9 + 1"], {:index=>2, :limit=>1})
[["\n\x00\x00\x00", "abc9 + 1", "\n\x00\x00\x00"]]

# box.select(["abc9 + 1", 10], {:index=>2, :limit=>1})
[["\n\x00\x00\x00", "abc9 + 1", "\n\x00\x00\x00"]]

# box.select(["abc9 + 2"], {:index=>1, :limit=>1})
[["\v\x00\x00\x00", "abc9 + 2", "\v\x00\x00\x00"]]

# box.select(["abc9 + 2"], {:index=>2, :limit=>1})
[["\v\x00\x00\x00", "abc9 + 2", "\v\x00\x00\x00"]]

# box.select(["abc9 + 2", 11], {:index=>2, :limit=>1})
[["\v\x00\x00\x00", "abc9 + 2", "\v\x00\x00\x00"]]

# box.select(9, 10, 11)
[["\t\x00\x00\x00", "abc9 + 0", "\t\x00\x00\x00"], ["\n\x00\x00\x00", "abc9 + 1", "\n\x00\x00\x00"], ["\v\x00\x00\x00", "abc9 + 2", "\v\x00\x00\x00"]]

# box.delete(9)
1

# box.select(9, 10, 11)
[["\n\x00\x00\x00", "abc9 + 1", "\n\x00\x00\x00"], ["\v\x00\x00\x00", "abc9 + 2", "\v\x00\x00\x00"]]

# box.delete(10)
1

# box.select(9, 10, 11)
[["\v\x00\x00\x00", "abc9 + 2", "\v\x00\x00\x00"]]

# box.delete(11)
1

# box.select(9, 10, 11)
[]

# box.select(0)
[]

# box.insert([0, "abc0", 0])
1

# box.select(0)
[["\x00\x00\x00\x00", "abc0", "\x00\x00\x00\x00"]]

# box.select(1)
[]

# box.insert([1, "abc1", 1])
1

# box.select(1)
[["\x01\x00\x00\x00", "abc1", "\x01\x00\x00\x00"]]

# box.select(2)
[]

# box.insert([2, "abc2", 2])
1

# box.select(2)
[["\x02\x00\x00\x00", "abc2", "\x02\x00\x00\x00"]]

# box.select(3)
[]

# box.insert([3, "abc3", 3])
1

# box.select(3)
[["\x03\x00\x00\x00", "abc3", "\x03\x00\x00\x00"]]

# box.select(4)
[]

# box.insert([4, "abc4", 4])
1

# box.select(4)
[["\x04\x00\x00\x00", "abc4", "\x04\x00\x00\x00"]]

# box.select(5)
[]

# box.insert([5, "abc5", 5])
1

# box.select(5)
[["\x05\x00\x00\x00", "abc5", "\x05\x00\x00\x00"]]

# box.select(6)
[]

# box.insert([6, "abc6", 6])
1

# box.select(6)
[["\x06\x00\x00\x00", "abc6", "\x06\x00\x00\x00"]]

# box.select(7)
[]

# box.insert([7, "abc7", 7])
1

# box.select(7)
[["\a\x00\x00\x00", "abc7", "\a\x00\x00\x00"]]

# box.select(8)
[]

# box.insert([8, "abc8", 8])
1

# box.select(8)
[["\b\x00\x00\x00", "abc8", "\b\x00\x00\x00"]]

# box.select(9)
[]

# box.insert([9, "abc9", 9])
1

# box.select(9)
[["\t\x00\x00\x00", "abc9", "\t\x00\x00\x00"]]

# box.select(0)
[["\x00\x00\x00\x00", "abc0", "\x00\x00\x00\x00"]]

# box.delete(0)
1

# box.select(0)
[]

# box.select(1)
[["\x01\x00\x00\x00", "abc1", "\x01\x00\x00\x00"]]

# box.delete(1)
1

# box.select(1)
[]

# box.select(2)
[["\x02\x00\x00\x00", "abc2", "\x02\x00\x00\x00"]]

# box.delete(2)
1

# box.select(2)
[]

# box.select(3)
[["\x03\x00\x00\x00", "abc3", "\x03\x00\x00\x00"]]

# box.delete(3)
1

# box.select(3)
[]

# box.select(4)
[["\x04\x00\x00\x00", "abc4", "\x04\x00\x00\x00"]]

# box.delete(4)
1

# box.select(4)
[]

# box.select(5)
[["\x05\x00\x00\x00", "abc5", "\x05\x00\x00\x00"]]

# box.delete(5)
1

# box.select(5)
[]

# box.select(6)
[["\x06\x00\x00\x00", "abc6", "\x06\x00\x00\x00"]]

# box.delete(6)
1

# box.select(6)
[]

# box.select(7)
[["\a\x00\x00\x00", "abc7", "\a\x00\x00\x00"]]

# box.delete(7)
1

# box.select(7)
[]

# box.select(8)
[["\b\x00\x00\x00", "abc8", "\b\x00\x00\x00"]]

# box.delete(8)
1

# box.select(8)
[]

# box.select(9)
[["\t\x00\x00\x00", "abc9", "\t\x00\x00\x00"]]

# box.delete(9)
1

# box.select(9)
[]

